File: android/src/main/java/io/acraig/capacitor/crypto/CryptoPlugin.java
package io.acraig.capacitor.crypto;

import android.os.Build;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "Crypto")
public class CryptoPlugin extends Plugin {

    private Crypto implementation = new Crypto();

    @PluginMethod
    public void echo(PluginCall call) {
        String value = call.getString("value");

        JSObject ret = new JSObject();
        ret.put("value", implementation.echo(value));
        call.resolve(ret);
    }

    @PluginMethod
    public void generateSharedSecret(PluginCall call)
    {
        String privateKey = call.getString("privateKey");
        String publicKey = call.getString("publicKey");

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            call.unavailable("Android Version too old!");
            return;
        }

        try
        {
            String sharedSecret = implementation.generateSharedSecret(publicKey, privateKey);

            JSObject ReturnValue = new JSObject();
            ReturnValue.put("sharedSecret", sharedSecret);

            call.resolve(ReturnValue);
        } catch (Exception e) {
            call.reject("Unable to generate Shared Secret", e);
        }
    }

    @PluginMethod
    public void generateRandomBytes(PluginCall call)
    {
        int byteCount = call.getInt("count");

        try
        {
            String randomBytesBase64 = implementation.generateRandomBytes(byteCount);

            JSObject ReturnValue = new JSObject();
            ReturnValue.put("bytesBase64", randomBytesBase64);

            call.resolve(ReturnValue);
        } catch (Exception e) {
            call.reject("Unable to generate random bytes", e);
        }
    }
    @PluginMethod
    public void generateKeyPair(PluginCall call)
    {
        JSObject ret = new JSObject();

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            call.unavailable("Android Version too old!");
            return;
        }

        try {
            String[] GeneratedKeys = implementation.generateKeyPair();

            JSObject ReturnValue = new JSObject();

            ReturnValue.put("privateKey", GeneratedKeys[Crypto.ARRAY_PRIVATE_KEY]);
            ReturnValue.put("publicKey", GeneratedKeys[Crypto.ARRAY_PUBLIC_KEY]);

            call.resolve(ReturnValue);
        } catch (Exception e) {
            call.reject("Unable to generate KeyPair: "+e.getMessage());
        }
    }

    @PluginMethod
    public void encrypt(PluginCall call) {
        JSObject ret = new JSObject();

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            call.unavailable("Android Version too old!");
            return;
        }

        String key = call.getString("key");
        String data = call.getString("data");
        try {
            String[] encryptRes = implementation.encrypt(key, data);

            ret.put("encryptedData", encryptRes[Crypto.ENCRYPT_CIPHERTEXT]);
            ret.put("tag", encryptRes[Crypto.ENCRYPT_TAG]);
            ret.put("iv", encryptRes[Crypto.ENCRYPT_IV]);

            call.resolve(ret);
        } catch (Exception e) {
            call.reject("Failed to encrypt", e);
        }
    }

    @PluginMethod
    public void decrypt(PluginCall call) {
        JSObject ret = new JSObject();

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            call.unavailable("Android Version too old!");
            return;
        }

        String key = call.getString("key");
        String encryptedData = call.getString("encryptedData");
        String tag = call.getString("tag");
        String iv = call.getString("iv");
        
        try {
            String decryptedData = implementation.decrypt(key, encryptedData, tag, iv);
            ret.put("unencryptedData", decryptedData);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("Failed to encrypt", e);
        }
    }
}

File: ios/Plugin/CryptoPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@available(iOS 14.0, *)
@objc(CryptoPlugin)
public class CryptoPlugin: CAPPlugin {
    private let implementation = Crypto();

    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        call.resolve([
            "value": implementation.echo(value)
        ])
    }
    
    @objc func generateKeyPair(_ call: CAPPluginCall) {
        let generatedKeyPair = implementation.generateKeyPair();
        call.resolve([
            "publicKey": generatedKeyPair.firstObject as! String,
            "privateKey": generatedKeyPair.lastObject as! String
        ])
    }
    
    @objc func generateRandomBytes(_ call: CAPPluginCall) {
        let bytesToGen = call.getInt("count") ?? 0;
        
        do {
            let randomBytes = try implementation.generateRandomBytes(bytesToGen);
            
            call.resolve([
                "bytesBase64": randomBytes
            ])
        } catch let error as NSError {
            call.reject("NSError: \(error.domain) \(error.localizedDescription) \(error.localizedFailureReason ?? "")");
        }
    }
    
    @objc func generateSharedSecret(_ call: CAPPluginCall) {
        do
        {
            let secretKey = try implementation.generateSharedSecret(call.getString("privateKey") ?? "", call.getString("publicKey") ?? "", call.getString("salt") ?? "")
            
            call.resolve([
                "sharedSecret": secretKey
            ]);
        } catch let error as NSError {
            call.reject("NSError: \(error.domain) \(error.localizedDescription) \(error.localizedFailureReason ?? "")");
        }
    }
    @objc func decrypt(_ call: CAPPluginCall) {
        do {
            guard let key = call.options["key"] as? String else {
                call.reject("Must provide a key")
                return
            }
            guard let encryptedData = call.options["encryptedData"] as? String else {
                call.reject("Must provide encryptedData")
                return
            }
            guard let tag = call.options["tag"] as? String else {
                call.reject("Must provide a tag")
                return
            }
            guard let iv = call.options["iv"] as? String else {
                call.reject("Must provide an iv")
                return
            }
            
            let res = try implementation.decrypt(key, encryptedData, tag, iv);

            call.resolve([
                "unencryptedData": res
            ]);
        } catch let error as NSError {
            call.reject("NSError: \(error.domain) \(error.localizedDescription) \(error.localizedFailureReason ?? "")");
        }
    }
    @objc func encrypt(_ call: CAPPluginCall)
    {
        do {
            guard let key = call.options["key"] as? String else {
                call.reject("Must provide a key")
                return
            }
            guard let data = call.options["data"] as? String else {
                call.reject("Must provide data")
                return
            }
            
            let res = try implementation.encrypt(key, data);

            call.resolve([
                "encryptedData": res[0],
                "tag": res[1],
                "iv": res[2]
            ]);
        } catch let error as NSError {
            call.reject("NSError: \(error.domain) \(error.localizedDescription) \(error.localizedFailureReason ?? "")");
        }
    }
    @objc func hash(_ call: CAPPluginCall) 
    {
        do {
            guard let data = call.options["data"] as? String else {
                call.reject("Must Provide Data");
                return;
            }
            
            let hash = try implementation.hash(data);
            call.resolve([
                "hash": hash
            ]);
        } catch let error as NSError {
            call.reject("NSError: \(error.domain) \(error.localizedDescription) \(error.localizedFailureReason ?? "")");
        }
    }
}
